### DISCUSSION TITLE
How to specify serial computation

### DISCUSSION THREAD

**Discussion by Ating24** [idaholab/moose#26729] [Q&A Tools]
I am using MOOSE's polynomial chaos for training on a cluster, calculating 256 deterministic samples. Using more kernels will keep initializing the equation, while using fewer kernels will exceed memory or cause the following errors：
  srun:First task exited 60s ago
  srun:StepId=128229.0 task 3:running
  srun :StepId=128229.0 tasks 0-2:exited
  srun:launch/slurm:step signal:Terminating StepId=128229.0
  srun :srun: Job step aborted: Waiting up to 32 seconds for job step to finish.130131
  srun: error: ion171:task 3:Killed
 when calculating the second step. Running the main input file alone can run it. Is there a way to abandon multiple apps and execute it in serial order? Or how to group the training sampler and specify the number of kernels for a single example

**Comment by GiudGiud** [comment]
Hello

You can limit the number of concurrent apps using the min/max procs per app parameters. 

If you still want only one processor to run with several assigned, then use a singleRankPartitioner in the Paritioning block to force it back to one processor

Guillaume

**Reply by Ating24** [reply]
Hello Under which module is it used? I tried using it and it seems to have no effect. Also, the MOOSE on the cluster is an early version, and there was no Reporters module at that time

**Reply by Ating24** [reply]
Does setting the number of threads have any impact on the task

**Reply by GiudGiud** [reply]
Using more threads will duplicate most moose objects but overall will cost less memory than distrivuted MPI parallelism because the mesh and the field variables won't be duplicated

**Reply by Ating24** [reply]
What is the relationship between the number of threads used and the number of kernels?

**Reply by Ating24** [reply]
I can run a single example in 1 hour with 80 kernels now. I want to run all 256 examples in series using 80 kernels. Do I set mode=batch - reset in the mutiapps module and then set min/max procs per app parameters in the samplers module

**Reply by Ating24** [reply]
For example, how many threads would be suitable for me to set using 80 kernels?

**Reply by GiudGiud** [reply]
> min/max procs per app

is in the MultiApps parameters. If you use a FullSolveSamplerMultiApp, you can see the parameters here
https://mooseframework.inl.gov/source/multiapps/SamplerFullSolveMultiApp.html

I think there's a misunderstanding on what a 'kernel' is here. What does it mean for you?

**Reply by grmnptr** [reply]
You have to specify a matching max/min-procs-per-row on the Sampler too.

**Reply by Ating24** [reply]
Is it like this?
```
[Samplers]
	[Quadrature]
		type = Quadrature
		order = 2 
		distributions = 'K_dist density_dist cp_dist fluxcoff_dist'
		execute_on- PRE_MULTIAPF_SETUP
		min_procs_per_row = 60
	[]
[]
[MultiApps]
	[roshan]
		type =SamplerFullSolveMultiApp
		input_files = file.i
		sampler =Quadraturemode - batch-reset
		min_procs_per_app = 60
	[]
[]
```

**Reply by Ating24** [reply]
But the warning displays：
```
pc_uniform.i:33: unused parameter 'Samplers/Quadrature/min_procs_per_row'
pc_uniform.i:33: unused parameter 'MultiApps/roshan/min_procs_per_app'
```

**Reply by Ating24** [reply]
This ‘kernel’ refers to the processor

**Reply by GiudGiud** [reply]
These parameters are odd?
```
execute_on- PRE_MULTIAPF_SETUP
		sampler =Quadraturemode - batch-reset
```
should be = not minus

PRE_MULTIAPP_SETUP is also misspelled

A kernel in moose is not a processor. It's an object computing a term in a PDE. It's best if you mention CPU or processor for a processor

**Reply by Ating24** [reply]
I took a look at the version of SamplerFullSolveMultiApp being used C did not mention min/max_ Procs_ Per_ App and min/max_ Procs_ Per_ Row

**Reply by Ating24** [reply]
The. i file is correct. I made an error copying it over

**Reply by GiudGiud** [reply]
Are you using a really old version of MOOSE?

**Reply by Ating24** [reply]
yes

**Reply by GiudGiud** [reply]
We're going to be pretty limited in getting an optimal parallel distribution then.
If you really care to run it as fast as possible, you should update

**Reply by Ating24** [reply]
So how can we calculate these 256 samples one by one or only calculate 2 or 4 samples at the same time

**Reply by Ating24** [reply]
At present, the update estimate is not very feasible. I just want to know if it is possible to specify how many processors to use for a single example or make it execute sequentially on the old version

**Reply by GiudGiud** [reply]
The parameters we mentioned will let you specify this for each child application.
You can specify however many processor on the main application with 
```
mpirun -n <number of procs> <executable_name-opt> -i <input_file.i>
```
but then the limitation of the number of processors in each child app can only be done with these parameters

**Reply by Ating24** [reply]
This only sets the number of processors used by the entire main program, but for each sub application, there is only one processor. I want to divide each processor into 80 processors, and then each sub application will use these 80 processes for calculation in sequence.

**Reply by GiudGiud** [reply]
Still, this can be done with the new parameters `min_procs_per`...

**Reply by Ating24** [reply]
What does it mean? Can the old version use this parameter?

**Reply by GiudGiud** [reply]
no. You have seen that this parameter is not used in the old version already

**Comment by grmnptr** [comment]
Also, make sure you use the batch-reset mode in combination with your min/max procs per app settings. Here is a short description on the gains:
https://mooseframework.inl.gov/modules/stochastic_tools/batch_mode.html 

The number of kernels would not influence your memory consumption too much unless you take it to the extreme.

**Reply by Ating24** [reply]
I can run a single example in 1 hour with 80 kernels now. I want to run all 256 examples in series using 80 kernels. Do I set mode=batch - reset in the mutiapps module and then set min/max procs per app parameters in the samplers module

**Reply by Ating24** [reply]
This version does not have the reprotters module, how to use min/max procs per app parameters

**Reply by Ating24** [reply]
I took a look at the version of SamplerFullSolveMultiApp being used C did not mention min/max_ Procs_ Per_ App and min/max_ Procs_ Per_ Row
